### ISSUE TITLE
create benchmarking script

### ISSUE THREAD

**Issue by IronicBadger** [trapexit/mergerfs#181]
Hi, 

I am trying to improve performance. I've used extensively AUFS and MHDDFS previously but MergerFS looks very promising. Presently with a 4gb file I achieve the following...

```
## Reading and writing to the array via mergerfs
# Write to array (new file, not overwrite)
root@epsilon:~# rsync --progress /tmp/test.file /mnt/storage/test.file
test.file
  6,087,566,826 100%   30.15MB/s    0:03:12 (xfr#1, to-chk=0/1)
# Read from array
root@epsilon:~# rsync --progress /mnt/storage/test.file /tmp/test.file
test.file
  6,087,566,826 100%   47.37MB/s    0:02:02 (xfr#1, to-chk=0/1)
```
I am reading from my 6 disk, all 3TB disk, array. If I read from the disk directly I achieve the following results...
```
## Direct disk access
# Write directly to disk
root@epsilon:~# rsync --progress /tmp/test.file /mnt/disk2/test.file
test.file
  6,087,566,826 100%  142.02MB/s    0:00:40 (xfr#1, to-chk=0/1)
# Read directly from disk
root@epsilon:~# rsync --progress /mnt/disk2/test.file /tmp/test.file
test.file
  6,087,566,826 100%   45.77MB/s    0:02:06 (xfr#1, to-chk=0/1)
```

From these outputs I can tell that writing to the array via MergerFS is much slower than I'd hoped, but that I have a bottleneck somewhere affecting reads directly from the disk. Any ideas where I'd start?

Hardware is more than capable of saturating the drives ( as demonstrated by the direct disk write ). Intel Avoton C2750 based platform with 16GB ECC RAM and the writes are being performed to an SSD with capabilities of 3-400MB/s writes. Data drives are formatted with XFS and SSD with ext4 on Debian 8. Please forgive my incredibly unscientific tests, it was all I could think of!

**Comment by trapexit** [comment]
There are a lot of variables to account for. I've been meaning to put together more thorough tests but haven't had the time. "Best case" would be mounting mergerfs over top tmpfs. In my experiences it was about half as fast but that's for a RAM disk so half that speed is still quite quick. Generally I see not drive saturating but certainly 1Gbit Ethernet saturating speeds as did Zack Reed in his tests[0].

Using `dd` with the proper arguments you can test transfer speeds pretty well. You can search Google for examples.

If you use `direct_io` you can increase write speeds at a loss of read which should be investigated too. You could mount them side by side too to compare more easily. Also, if you are using an older kernel or version of FUSE you'll get much slower speed on account of not using `splice`. If you let me know what you kernel and fuse versions are I can comment further. Make sure to have `defaults` set to get the general optimal settings.

If there is a lot of contention (many open files reading and writing or being opened/stat'ed) I find libfuse doesn't scale terribly well. I'm going to be working with them to improve that. Also a feature is being considered which would allow reads and write to be almost native speed which should help quite a bit but even if accepted it'd take a while for it to be available.

[0] http://zackreed.me/articles/92-mergerfs-another-good-option-to-pool-your-snapraid-disks

**Comment by IronicBadger** [comment]
Thanks for your reply... I decided to reinstall MHDDFS and take a look at the difference in performance. To do this I ended up installing a patched version from (here)[https://github.com/vdudouyt/mhddfs-nosegfault] - which meant I enabled Jessie unstable repos to get a newer version of `libfuse-dev`. Consequently I am getting better performance on both MHDDFS *and* MergerFS now, as the results below show. I conducted each test after the repos were enabled 3 times and posted the result in the middle speed wise.

Presumably enabling the unstable repos upgraded a few things in the background which have helped my performance greatly. Feel free to close this issue, I post merely to help others in future. I'll also be doing a write up for linuxserver.io, my site, in the next few days. 

**MergerFS** - before unstable enabled
```
root@epsilon:~# dd if=/dev/zero of=/mnt/storage/test.img bs=1M count=10000
10000+0 records in
10000+0 records out
10485760000 bytes (10 GB) copied, 189.048 s, 55.5 MB/s
```

**MergerFS** - after unstable repos enabled
```
root@epsilon:~# dd if=/dev/zero of=/mnt/storage/test.img bs=512M count=50
50+0 records in
50+0 records out
26843545600 bytes (27 GB) copied, 278.909 s, 96.2 MB/s
root@epsilon:~# rm /mnt/storage/test.img
root@epsilon:~# dd if=/dev/zero of=/mnt/storage/test.img bs=1M count=10000
10000+0 records in
10000+0 records out
10485760000 bytes (10 GB) copied, 124.601 s, 84.2 MB/s

```

**MHDDFS** - after unstable repos enabled
```
root@epsilon:~# dd if=/dev/zero of=/mnt/storage/test.img bs=512M count=50
50+0 records in
50+0 records out
26843545600 bytes (27 GB) copied, 292.922 s, 91.6 MB/s
root@epsilon:~# rm /mnt/storage/test.img 
root@epsilon:~# dd if=/dev/zero of=/mnt/storage/test.img bs=1M count=10000
10000+0 records in
10000+0 records out
10485760000 bytes (10 GB) copied, 118.769 s, 88.3 MB/s
```

**Comment by trapexit** [comment]
The speeds shouldn't be too different though mergerfs is far more thorough in terms of security and such so certain behaviors can be slower probably. For pure read and writes I'd expect mergerfs to be slightly faster.

Be sure to test with direct_io on and off and you should be sure to use sync and non-sync'ed reads/writes[0] and force the cache to clear between each test[1].

As I mentioned I'll be creating a simple shell script to run multiple permutations of these so it's easier to do these tests in the future.

[0] https://romanrm.net/dd-benchmark
[1] http://unix.stackexchange.com/questions/150672/dd-write-vs-read-performance

**Comment by trapexit** [comment]
Also... the FUSE settings can make massive differences. Make sure mhddfs and mergerfs are mounted with the same options. You can read about what `defaults` means in the mergerfs docs. Ultimately any benchmark script will need to takes those into account as well. It's certainly possible that my `defaults` settings could be optimized.

**Comment by rubylaser** [comment]
@IronicBadger I thought you had went back to UnRAID?  Are you just testing the waters :)  I've been sad to see the forum on your site slow down in terms of traffic.  There were some good discussions going on there.  Your dd tests on the local host I would think should be faster (I wouldn't use a 512MB block size as that's bit bigger than what a standard write blocksize would be (stick to 1 or 4MB)... Also, try to go with a filesize at least as a big as your RAM (more like double) and sync between writing and reading.

Here's a test on my server with an i5-4590 and 16GB of RAM.  Note the > than 2x my RAM at 34GB test size and syncing between tests with 1MB blocksizes (Ubuntu 14.04.3 Server).  This was all to a Seagate 4TB Desktop drive, so not one of the faster disks in my pool  (I have a bunch of 6TB WD Red drives, that are much faster).

**Write**
```
root@fileserver:~# dd if=/dev/zero of=/storage/testfile.out bs=1M count=32000
32000+0 records in
32000+0 records out
33554432000 bytes (34 GB) copied, 211.511 s, 159 MB/s
root@fileserver:~# sync
```

**Read**
```
root@fileserver:~# dd if=/storage/testfile.out of=/dev/null bs=1M
32000+0 records in
32000+0 records out
33554432000 bytes (34 GB) copied, 190.572 s, 176 MB/s
```

**Comment by IronicBadger** [comment]
hey @rubylaser - i went back to unRAID for all of a month before remembering why I left in the first place (lack of flexibility in the base OS mainly). I switched out to ZFS mirrored pairs but ran out of space quite quickly and ended up moving back once more to a snapraid / pooled solution after seeing your post on mergerfs.

**Comment by rubylaser** [comment]
@IronicBadger Well, I'm glad you are back in the fold.  UnRAID is a nice solution for some, it's just not the one for me.

**Comment by IronicBadger** [comment]
OK. Ubuntu 14.04.3 with fuse-2.9.4 (install from wily src)

Here are the lines from `/etc/fstab` which matter.
```
## MergerFS
/mnt/disk*  /mnt/storage  fuse.mergerfs  defaults,allow_other,minfreespace=50G,fsname=mergerfs 0 0

## MHDDFS
/mnt/disk1,/mnt/disk2,/mnt/disk3,/mnt/disk4,/mnt/disk5 /mnt/storage fuse.mhddfs defaults,allow_other,nonempty,mlimit=50G,fsname=mhddfs 0 0
```

Both test files were written to the same physical drive (disk4), and just for my sanity I did a raw write (bypassing FUSE) to the drive before starting. I flushed the cache everytime between writes and reads as suggested on your links @trapexit to make it as fairer test as possible.

## Raw disk4 write
```
# write
root@epsilon:~# dd bs=1M count=16000 if=/dev/zero of=/mnt/disk4/test.disk4 conv=fdatasync
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 110.93 s, 151 MB/s

# read
root@epsilon:~# dd if=/mnt/disk4/test.disk4 of=/dev/null bs=1M
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 107.612 s, 156 MB/s
```

## MergerFS
```
# write
dd bs=1M count=16000 if=/dev/zero of=/mnt/storage/test.merger conv=fdatasync
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 305.269 s, 55.0 MB/s

# read
root@epsilon:/home/alex# dd if=/mnt/storage/test.merger of=/dev/null bs=1M
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 107.624 s, 156 MB/s
```

## MHDDFS (with nosegfault patch)
```
# write
root@epsilon:/home/alex# dd bs=1M count=16000 if=/dev/zero of=/mnt/storage/test.mhddfs conv=fdatasync
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 298.938 s, 56.1 MB/s

# read
root@epsilon:~# dd if=/mnt/storage/test.mhddfs of=/dev/null bs=1M
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 107.666 s, 156 MB/s
```

@rubylaser I have more powerful hardware than you, disks which are just as fast and yet I can't come anywhere close to your speeds. I thought upgrading the fuse version from 2.9.2 to 2.9.4 would make the difference, but it hasn't. I'd love to know which package in the debian unstable test I did last week made the difference - 56 MB/s write just isn't good enough!

**Comment by rubylaser** [comment]
@IronicBadger I would agree that 56 MB/s is too slow especially for a local test (I see gigabit speeds over Samba, to my pool, so this should be almost full disk speed locally). 

As an FYI,  I have an i5-4590 vs. your C2750.  My CPU has less physical cores than yours, but has much higher base and boost clock speeds than yours.  Also, it has almost twice the performance overall and 4 times the performance in single threaded applications.  I wonder if that's exacerbating your issue.  Here's a Passmark comparing the two CPUs.

[i5-4590 vs. C2750](https://www.cpubenchmark.net/compare.php?cmp[]=2234&cmp[]=2185)

Also, I'm not using the latest FUSE on my Ubuntu 14.04.3 machine.  If you post the directions you used to compile the latest FUSE, I'd be wiling to compare performance on my machine to see if it changes anything.

Also, I tried again to see if forcing the sync with conv=fdatasync as part of the dd process rather than syncing afterward made mergerfs slower, but I see exactly the same numbers either way.

```
# Write
root@fileserver:~# dd if=/dev/zero of=/storage/testfile.out bs=1M count=32000 conv=fdatasync
32000+0 records in
32000+0 records out
33554432000 bytes (34 GB) copied, 205.394 s, 163 MB/s

# Read
root@fileserver:~# dd if=/storage/testfile.out of=/dev/null bs=1M
32000+0 records in
32000+0 records out
33554432000 bytes (34 GB) copied, 207.431 s, 162 MB/s
```

**Comment by IronicBadger** [comment]
Oh maybe it is the CPU, but I don't think so...

I replaced `trusty` with `wily` e.g. `deb-src http://gb.archive.ubuntu.com/ubuntu/ wily main restricted` in my `/etc/apt/sources.list` file. Then I installed `apt-src` and did something like `apt-src --build install libfuse2`. That's from memory though...

**Comment by trapexit** [comment]
Have you tried using `direct_io` and / or setting the splice arguments individually?

**Comment by IronicBadger** [comment]
I'm using the ubuntu 14.04.3 defaults, whatever they are with regard to `direct_io` and haven't touched any splice settings either.

My kernel is `Linux epsilon 3.19.0-42-generic #48~14.04.1-Ubuntu SMP Fri Dec 18 10:24:49 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by trapexit** [comment]
`direct_io` and splice are FUSE options. Look at the mergerfs man page / readme and `mergerfs -h`. 

Try the test with and without `direct_io` and then replace `defaults` with `splice_read`, `splice_write`, and/or `splice_move`.

**Comment by IronicBadger** [comment]
With `direct_io` I got the following speeds:

```
# write
root@epsilon:~# dd bs=1M count=16000 if=/dev/zero of=/mnt/disk4/test.mergerfs conv=fdatasync
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 115.137 s, 146 MB/s

# read
root@epsilon:~# dd if=/mnt/disk4/test.mergerfs of=/dev/null bs=1M
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 107.739 s, 156 MB/s
```

**Comment by IronicBadger** [comment]
I tried a few different combinations and the results are below...

```
/mnt/disk*  /mnt/storage  fuse.mergerfs  direct_io,defaults,allow_other,minfreespace=50G,fsname=mergerfs 0 0

# write
root@epsilon:~# dd bs=1M count=16000 if=/dev/zero of=/mnt/storage/test.mergerfs conv=fdatasync
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 124.345 s, 135 MB/s

# read
root@epsilon:~# dd if=/mnt/storage/test.mergerfs of=/dev/null bs=1M
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 107.847 s, 156 MB/s
```

And now with the splices...
```
/mnt/disk*  /mnt/storage  fuse.mergerfs  direct_io,splice_read,splice_write,allow_other,minfreespace=50G,fsname=mergerfs 0 0

# write
root@epsilon:~# dd bs=1M count=16000 if=/dev/zero of=/mnt/storage/test.mergerfs conv=fdatasync
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 125.088 s, 134 MB/s

# read
root@epsilon:~# dd if=/mnt/storage/test.mergerfs of=/dev/null bs=1M
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 108.009 s, 155 MB/s
```

And now without `direct_io` but with `splice_read` and `splice_write` it drops back down in speed.
```
/mnt/disk*  /mnt/storage  fuse.mergerfs  splice_read,splice_write,allow_other,minfreespace=50G,fsname=mergerfs 0 0

# write
root@epsilon:~# dd bs=1M count=16000 if=/dev/zero of=/mnt/storage/test.mergerfs conv=fdatasync
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 325.336 s, 51.6 MB/s

# read
root@epsilon:~# dd if=/mnt/storage/test.mergerfs of=/dev/null bs=1M
16000+0 records in
16000+0 records out
16777216000 bytes (17 GB) copied, 107.971 s, 155 MB/s
```

**TL;DR** Just enable `direct_io` in your fstab.

@trapexit are there any downsides to using `direct_io` on permanently?

**Comment by trapexit** [comment]
Shouldn't be an issue leaving direct_io enabled. It just disables some caching. Cached reads may be slower but you probably won't notice.

**Comment by rubylaser** [comment]
@IronicBadger I'm glad you got it figured out!  Those are much better speeds.  I'm half tempted to try direct_io myself, but I'm happy with my current performance, so I'd better leave well enough alone :)

**Comment by trapexit** [comment]
Moved to https://github.com/trapexit/mergerfs-tools/issues/4
